File: src/main/kotlin/com/linqingying/cangjie/psi/CjTypeParameter.java
package com.linqingying.cangjie.psi;

import com.linqingying.cangjie.lexer.CjTokens;
import com.linqingying.cangjie.psi.stubs.CangJieTypeParameterStub;

import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import com.intellij.psi.search.LocalSearchScope;
import com.intellij.psi.search.SearchScope;
import com.intellij.psi.util.PsiTreeUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import com.linqingying.cangjie.psi.stubs.elements.CjStubElementTypes;

public class CjTypeParameter extends CjNamedDeclarationStub<CangJieTypeParameterStub> {

    public CjTypeParameter(@NotNull ASTNode node) {
        super(node);
    }

    public CjTypeParameter(@NotNull CangJieTypeParameterStub stub) {
        super(stub, CjStubElementTypes.TYPE_PARAMETER);
    }

    @Override
    public <R, D> R accept(@NotNull CjVisitor<R, D> visitor, @Nullable D data) {
        return visitor.visitTypeParameter(this, data);
    }


    @Override
    public String toString() {
        return  getNode().getElementType().toString();
    }


    @Nullable
    public CjTypeReference setExtendsBound(@Nullable CjTypeReference typeReference) {
        CjTypeReference currentExtendsBound = getExtendsBound();
        if (currentExtendsBound != null) {
            if (typeReference == null) {
                PsiElement colon = findChildByType(CjTokens.COLON);
                if (colon != null) colon.delete();
                currentExtendsBound.delete();
                return null;
            }
            return (CjTypeReference) currentExtendsBound.replace(typeReference);
        }

        if (typeReference != null) {
            PsiElement colon = addAfter(new CjPsiFactory(getProject()).createColon(), getNameIdentifier());
            return (CjTypeReference) addAfter(typeReference, colon);
        }

        return null;
    }

    @Nullable
    public CjTypeReference getExtendsBound() {
        return getStubOrPsiChild(CjStubElementTypes.TYPE_REFERENCE);
    }

    @NotNull
    @Override
    public SearchScope getUseScope() {
        CjTypeParameterListOwner owner = PsiTreeUtil.getParentOfType(this, CjTypeParameterListOwner.class);
        return new LocalSearchScope(owner != null ? owner : this);
    }
}

File: src/main/kotlin/com/linqingying/cangjie/psi/CjTypeParameter.kt
package com.linqingying.cangjie.psi

import com.linqingying.cangjie.lexer.CjTokens
import com.linqingying.cangjie.psi.stubs.CangJieTypeParameterStub
import com.linqingying.cangjie.psi.stubs.elements.CjStubElementTypes
import com.linqingying.cangjie.types.Variance
import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElement
import com.intellij.psi.search.LocalSearchScope
import com.intellij.psi.search.SearchScope
import com.intellij.psi.util.PsiTreeUtil

class CjTypeParameter : CjNamedDeclarationStub<CangJieTypeParameterStub > {
    constructor(node: ASTNode) : super(node)

    constructor(stub: CangJieTypeParameterStub) : super(stub, CjStubElementTypes.TYPE_PARAMETER)

    override fun <R, D> accept(visitor: CjVisitor<R, D>, data: D?): R {
        return visitor.visitTypeParameter(this, data)
    }


    override fun toString(): String {
        return node.elementType.toString()
    }

    val variance: Variance
        get() =//        CangJieTypeParameterStub stub = getStub();
//        if (stub != null) {
//
////            if (stub.isInVariance()) return Variance.IN_VARIANCE;
//            return Variance.INVARIANT;
//        }
//
//        CjModifierList modifierList = getModifierList();
//        if (modifierList == null) return Variance.INVARIANT;
            Variance.INVARIANT

    fun setExtendsBound(typeReference: CjTypeReference?): CjTypeReference? {
        val currentExtendsBound = extendsBound
        if (currentExtendsBound != null) {
            if (typeReference == null) {
                val colon = findChildByType<PsiElement>(CjTokens.COLON)
                colon?.delete()
                currentExtendsBound.delete()
                return null
            }
            return currentExtendsBound.replace(typeReference) as CjTypeReference
        }

        if (typeReference != null) {
            val colon = addAfter(CjPsiFactory(project).createColon(), nameIdentifier)
            return addAfter(typeReference, colon) as CjTypeReference
        }

        return null
    }

    val extendsBound: CjTypeReference?
        get() = getStubOrPsiChild(CjStubElementTypes.TYPE_REFERENCE)
    val extendsBounds: List<CjTypeReference>
        get() = getStubOrPsiChildrenAsList(
            CjStubElementTypes.TYPE_REFERENCE
        )

    override fun getUseScope(): SearchScope {
        val owner = PsiTreeUtil.getParentOfType(
            this,
            CjTypeParameterListOwner::class.java
        )
        return LocalSearchScope(owner ?: this)
    }
}
